src/nano_zlib.cc

```diff
@@ -45,18 +45,128 @@ word ZlibRle::add(const uint8* contents, intptr_t extra) {
       return i;
     }
     uint8 b = contents[i];
-    if (b == last_byte_) {
-      repetitions_++;
-      // We can max output 257 repetitions.  If the number gets too high then
-      // flush a bit to limit how much data we will want to output later, but
-      // don't leave less than 3, because there are no codes for a repetition
-      // of 1 or 2 and it has to be done in a more verbose way.
-      if (repetitions_ >= 260) output_repetitions();
+    if (mode_ == LITERAL) {
+      // We need at least 4 bytes of lookahead to make compression decisions.
+      if (unemitted_bytes_valid_ < 4) {
+        unemitted_bytes_ <<= 8;
+        unemitted_bytes_ |= b;
+        unemitted_bytes_valid_++;
+        continue;
+      }
+      if (last_bytes_valid_ >= 1 &&
+          (last_bytes_ & 0xff) == (unemitted_bytes_ & 0xff) &&
+          (last_bytes_ & 0xff) == ((unemitted_bytes_ >> 8) & 0xff) &&
+          (unemitted_bytes_ >> 16) == (unemitted_bytes_ & 0xffff)) {
+        // We recognized that the unemitted bytes all four matched the last
+        // emitted one.
+        mode_ = REP1;
+        repetitions_ = 4;
+        last_bytes_ = unemitted_bytes_;
+        last_bytes_valid_ = 4;
+        unemitted_bytes_ = 0;
+        unemitted_bytes_valid_ = 0;
+      } else if (last_bytes_valid_ >= 2 &&
+          (last_bytes_ & 0xffff) == (unemitted_bytes_ & 0xffff) &&
+          (unemitted_bytes_ >> 16) == (unemitted_bytes_ & 0xffff)) {
+        // We recognized that the four unemitted bytes are two repetitions of
+        // the last two emitted ones.
+        mode_ = REP2;
+        repetitions_ = 4;
+        last_bytes_ = unemitted_bytes_;
+        last_bytes_valid_ = 4;
+        unemitted_bytes_ = 0;
+        unemitted_bytes_valid_ = 0;
+      } else if (last_bytes_valid_ >= 1 &&
+          (last_bytes_ & 0xff) == ((unemitted_bytes_ >> 8) & 0xff) &&
+          (last_bytes_ & 0xff) == ((unemitted_bytes_ >> 16) & 0xff) &&
+          (last_bytes_ & 0xff) == ((unemitted_bytes_ >> 24) & 0xff)) {
+        // We recognized that the three first unemitted bytes are repetitions
+        // of the last emitted one.  It's not a huge win to code them as
+        // a run of 1+3, but it saves a few bits.
+        mode_ = REP1;
+        repetitions_ = 3;
+        last_bytes_ <<= 24;
+        last_bytes_ |= unemitted_bytes_ >> 8;
+        last_bytes_valid_ = 4;
+        unemitted_bytes_ &= 0xff;
+        unemitted_bytes_valid_ = 1;
+        output_repetitions();  // Will flush completely since repetitions_ == 3.
+      } else if (last_bytes_valid_ >= 4 &&
+          last_bytes_ == unemitted_bytes_) {
+        // We recognized that the four unemitted bytes are the same as
+        // the last four emitted ones.  Start a run with offset 4.
+        mode_ = REP4;
+        repetitions_ = 4;
```

Comment:
Same: it's just one repetition of 4 bytes. Not 4 of them.

Reply:
Renamed to bytes_repeated_